rfc-087-dealing-with-errors.md

```diff
@@ -0,0 +1,105 @@
+# Dealing with errors
+
+## Summary
+
+This describes how we treat errors on GOV.UK
+
+## Problem
+
+We've recently migrated to a new error tracking service, Sentry. This provides us an opportunity to rethink how we treat errors.
+
+## Proposal
+
+There are 2 principles:
+
+### 1. When something goes wrong, we should be notified
+
+Applications should report exceptions to Sentry. Applications must not swallow errors.
+
+### 2. Applications should not error
+
+The goal of GOV.UK is that applications should not error. When something goes wrong it should be fixed.
+
+## Classifying errors
+
+### Bug
+
+A code change makes the application crash.
+
+Desired behaviour: error is sent to Sentry, developers are notified and fix the error. Developers mark the error in Sentry as `Resolved`. This means a recurrence of the error will alert developers again.
+
+### Intermittent errors without user impact
+
+Frontend applications often see timeouts when talking to the content-store.
+
+There's no or little user impact because the request will be answered by the caching layer.
+
+Example: <https://sentry.io/govuk/app-finder-frontend/issues/352985400>
+
+Desired behaviour: error is not sent to Sentry. Instead, we rely on Smokey and Icinga checks to make sure we the site functions.
+
+### Intermittent errors with user impact
+
+Publishing applications sometimes see timeouts when talking to publishing-api. This results in the publisher seeing an error page and possibly losing data.
+
+Example: <https://sentry.io/govuk/app-content-tagger/issues/367277928>
+
+Desired behaviour: TBD
+
+### Intermittent retryable errors
+
+Sidekiq worker sends something to the publishing-api, which times out. Sidekiq retries, the next time it works.
+
+Desired behaviour: errors are not reported to Sentry until retries are exhausted. See [this PR for an example](https://github.com/alphagov/content-performance-manager/pull/353).
+
+Relevant: https://github.com/getsentry/raven-ruby/pull/784
+
+### Expected environment-based errors
+
+MySQL errors on staging while data sync happens.
+
+Example: <https://sentry.io/govuk/app-whitehall/issues/343619055>
+
+Desired behaviour: our environment is set up such that these errors do not occur.
+
+### Bad request errors
+
+User makes a request the application can't handle ([example][bad-request]).
+
+Often happens in [security checks](https://sentry.io/govuk/app-frontend/issues/400074979).
+
+Example: <https://sentry.io/govuk/app-frontend/issues/400074979>
+
+Desired behaviour: user gets feedback, error is not reported to Sentry
+
+[bad-request]: https://sentry.io/govuk/app-service-manual-frontend/issues/400074003
+
+### Incorrect bubbling up of errors
```

Comment:
Rummager is absolutely doing the wrong thing in this case. The error should be returned as 4xx, and the upstream app should either deal with it or not generate invalid requests. But I think there is another issue here...
 
If an upstream app can't do something because of a failure in a dependent service, that is a critical piece of information to have in the application logs. By not reporting this as an error with the upstream app, I am more likely to see that the actual broken app is broken. But I still might want to investigate what the impact was on upstream applications, in which case I want to see something in kibana. On the other hand if I start investigating a problem by looking at the upstream app, seeing that log message signposts the root cause and might prompt me to go look at sentry for that app in the first place.

So I think this is another case where we should not take the information out of sentry without first making sure we have good metrics or logging in other places.

Reply:
Is this the sort of thing where we can tag the error with something in sentry so it can be isolated?

Reply:
I think I had this one wrong. Rummager does the right thing by returning a `422`. Fixed in https://github.com/alphagov/govuk-rfcs/pull/87/commits/f78767c0d339fdcdde0083b5537784a6ad81d566.